src/BSON/Int64.c

```diff
@@ -244,6 +263,100 @@ static int php_phongo_int64_compare_objects(zval* o1, zval* o2)
 	return 0;
 }
 
+static zend_result php_phongo_int64_cast_object(phongo_compat_object_handler_type* readobj, zval* retval, int type)
+{
+	php_phongo_int64_t* intern;
+
+	intern = Z_OBJ_INT64(PHONGO_COMPAT_GET_OBJ(readobj));
+
+	switch (type) {
+		case IS_LONG:
+#if PHP_VERSION_ID >= 70300
+		case _IS_NUMBER:
+#endif
+#if SIZEOF_ZEND_LONG == 4
+			if (intern->integer > INT32_MAX || intern->integer < INT32_MIN) {
+				zend_error(E_WARNING, "Truncating 64-bit integer value %" PRId64 " to 32 bits", intern->integer);
+			}
+#endif
+
+			ZVAL_LONG(retval, intern->integer);
+
+			return SUCCESS;
+
+		case _IS_BOOL:
+			ZVAL_BOOL(retval, intern->integer != 0);
+
+			return SUCCESS;
+
+		default:
+			return zend_std_cast_object_tostring(readobj, retval, type);
+	}
+}
+
+static zend_result php_phongo_int64_do_operation(zend_uchar opcode, zval* result, zval* op1, zval* op2)
+{
+	int64_t value1, value2;
+
+	if (Z_TYPE_P(op1) == IS_LONG) {
+		value1 = Z_LVAL_P(op1);
+	} else if (Z_TYPE_P(op1) == IS_OBJECT && Z_OBJCE_P(op1) == php_phongo_int64_ce) {
+		value1 = Z_INT64_OBJ_P(op1)->integer;
+	} else {
+		return FAILURE;
+	}
+
+	if (Z_TYPE_P(op2) == IS_LONG) {
+		value2 = Z_LVAL_P(op2);
+	} else if (Z_TYPE_P(op2) == IS_OBJECT && Z_OBJCE_P(op2) == php_phongo_int64_ce) {
+		value2 = Z_INT64_OBJ_P(op2)->integer;
+	} else {
+		return FAILURE;
+	}
+
+	switch (opcode) {
+		case ZEND_ADD:
+			ZVAL_INT64_OBJ(result, value1 + value2);
+			return SUCCESS;
+
+		case ZEND_SUB:
+			ZVAL_INT64_OBJ(result, value1 - value2);
+			return SUCCESS;
+
+		case ZEND_MUL:
+			ZVAL_INT64_OBJ(result, value1 * value2);
+			return SUCCESS;
+
+		case ZEND_DIV:
+			if (value2 == 0) {
+				zend_throw_exception(zend_ce_division_by_zero_error, "Division by zero", 0);
+				return FAILURE;
+			}
+
+			ZVAL_INT64_OBJ(result, value1 / value2);
+			return SUCCESS;
+
+		case ZEND_MOD:
+			ZVAL_INT64_OBJ(result, value1 % value2);
+			return SUCCESS;
+
+		case ZEND_SL:
+			ZVAL_INT64_OBJ(result, value1 << value2);
+			return SUCCESS;
+
+		case ZEND_SR:
+			ZVAL_INT64_OBJ(result, value1 >> value2);
+			return SUCCESS;
+
+		case ZEND_POW:
+			ZVAL_INT64_OBJ(result, pow(value1, value2));
+			return SUCCESS;
+
+		default:
```

Comment:
* `ZEND_BOOL_NOT` and `ZEND_BOOL_XOR` worked through implicitly casting, and I added tests for that.
* Increment/decrement is converted to `ZEND_ADD`/`ZEND_SUB`, but tests were added
* Bitwise operations were added
* `fast_long_add_function` (and all other built-in functions) can't be used due to them using `zend_long`, which on 32-bit platforms is `long` instead of `long long`. If we ever drop 32-bit support we can revisit this, but for now we can't leverage them.